File: source/aoBlogs2/Views/SidebarView.vb

Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports System.Text.RegularExpressions
Imports Contensive.Addons.Blog.Controllers
Imports Contensive.Addons.Blog.Models
Imports Contensive.Addons.Blog.Models.View
Imports Contensive.BaseClasses

Namespace Views
    '
    Public Class SidebarView
        '
        Public Shared Function getSidebarView(cp As CPBaseClass, app As ApplicationEnvironmentModel, request As RequestModel, legacyBlogBody As String) As String
            Try
                Dim blog As BlogModel = app.blog
                Dim cellList As String = ""
                Using sidebarCell As CPBlockBaseClass = cp.BlockNew()
                    Dim layout As CPBlockBaseClass = cp.BlockNew()
                    layout.OpenLayout(BlogListLayout)
                    sidebarCell.Load(layout.GetOuter(".blogSidebarCellWrap"))
                    Dim cellTemplate As String = sidebarCell.GetHtml()
                    Dim adminSuggestions As String = ""
                    Dim sidebarCnt As Integer = 0
                    '
                    ' -- wrap the blog body in the wrapper (sidebar, footer, etc)
                    Dim isArticleView As Boolean = (request.blogEntryId > 0)
                    Dim allowListSidebar As Boolean = Blog.allowEmailSubscribe Or Blog.allowFacebookLink Or Blog.allowGooglePlusLink Or Blog.allowGooglePlusLink Or Blog.allowRSSSubscribe Or Blog.allowTwitterLink Or Blog.allowArchiveList Or Blog.allowSearch
                    Dim allowArticleSidebar As Boolean = allowListSidebar Or Blog.allowArticleCTA
                    Dim dstFormId As Integer = cp.Doc.GetInteger(RequestNameFormID)
                    Dim allowSidebar As Boolean = Not (dstFormId = FormBlogEntryEditor) And ((isArticleView And allowArticleSidebar) Or (Not isArticleView And allowListSidebar))
                    '
                    ' Sidebar
                    If allowSidebar Then
                        If Blog.allowArticleCTA And isArticleView Then
                            '
                            ' CTA cells
                            Dim blogEntryCtaRuleList = DbModel.createList(Of BlogEntryCTARuleModel)(cp, "blogEntryid=" & request.blogEntryId)
                            For Each rule In blogEntryCtaRuleList
                                Dim cta = DbModel.create(Of CallsToActionModel)(cp, rule.calltoactionid)
                                If (cta IsNot Nothing) Then
                                    sidebarCell.Load(cellTemplate)
                                    Call sidebarCell.SetInner(".blogSidebarCellHeadline", cta.headline & "<br>")
                                    Call sidebarCell.SetInner(".blogSidebarCellCopy", cta.brief)
                                    Call sidebarCell.SetOuter(".blogSidebarCellInput", "")
                                    Call sidebarCell.SetOuter(".blogSidebarCellInputCaption", "")
                                    If (cta.name = "") Or (cta.link = "") Then
                                        Call sidebarCell.SetOuter(".blogSidebarCellButton", "")
                                    Else
                                        Call sidebarCell.SetInner(".blogSidebarCellButton", "<a target=""_blank"" href=""" & cta.link & """>" & cta.name & "</a>")
                                    End If
                                    cellList &= vbCrLf & vbTab & sidebarCell.GetHtml()
                                    sidebarCnt += 1
                                End If
                            Next
                        End If
                        '
                        If Blog.allowEmailSubscribe Then
                            '
                            ' Subscribe by email
                            '
                            Dim subscribed As Boolean = cp.Visit.GetBoolean("EmailSubscribed-Blog" & Blog.id & "-user" & cp.User.Id)
                            If Not subscribed Then
                                subscribed = cp.User.IsInGroup(Blog.emailSubscribeGroupId.ToString())
                            End If
                            sidebarCell.Load(cellTemplate)
                            Call sidebarCell.SetInner(".blogSidebarCellHeadline", "Subscribe By Email")
                            Call sidebarCell.SetOuter(".blogSidebarCellCopy", "")
                            If subscribed Then
                                Call sidebarCell.SetInner(".blogSidebarCellInputCaption", "You are subscribed to this blog.")
                                Call sidebarCell.SetOuter(".blogSidebarCellInput", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellButton", "")

                            Else
                                Call sidebarCell.SetInner(".blogSidebarCellInputCaption", "Email*")
                                Call sidebarCell.SetInner(".blogSidebarCellInput", "<input type=""text"" id=""blogSubscribeEmail"" name=""email"" value=""" & cp.User.Email & """>")
                                Call sidebarCell.SetInner(".blogSidebarCellButton a", "Subscribe")
                                Call sidebarCell.SetInner(".blogSidebarCellButton", "<a href=""#"" id=""blogSidebarEmailSubscribe"">Subscribe</a>")
                            End If
                            cellList &= vbCrLf & vbTab & "<div id=""blogSidebarEmailCell"">" & sidebarCell.GetHtml() & "</div>"
                            sidebarCnt += 1
                        End If
                        '
                        If Blog.allowFacebookLink Or Blog.allowGooglePlusLink Or Blog.allowTwitterLink Then
                            '
                            ' Social Links
                            'Dim sidebarHtml As String = sidebarCell.GetHtml()
                            Dim sidebarHtml As String = ""
                            If Blog.allowFacebookLink And (Blog.facebookLink <> "") Then
                                sidebarHtml &= "<a href=""" & Blog.facebookLink & """ target=""_blank""><img Class=""blogSidebarSocialLogo"" src=""/blogs/facebook.jpg"" width=""32"" height=""32""></a>"
                            ElseIf Blog.allowFacebookLink Then
                                If cp.User.IsAdmin Then
                                    sidebarHtml &= "<div Class=""blogAdminWarning""><h2>Administrator</h2><p>Add a facebook link For this blog, Or disable the Allow Facebook Sidebar checkbox.</p></div>"
                                End If
                            End If
                            If Blog.allowTwitterLink And (Blog.twitterLink <> "") Then
                                sidebarHtml &= "<a href=""" & Blog.twitterLink & """ target=""_blank""><img Class=""blogSidebarSocialLogo"" src=""/blogs/twitter.jpg"" width=""32"" height=""32""></a>"
                            ElseIf Blog.allowTwitterLink Then
                                If cp.User.IsAdmin Then
                                    sidebarHtml &= "<div Class=""blogAdminWarning""><h2>Administrator</h2><p>Add a twitter link For this blog, Or disable the Allow Twitter Sidebar checkbox.</p></div>"
                                End If
                            End If
                            If Blog.allowTwitterLink And (Blog.googlePlusLink <> "") Then
                                sidebarHtml &= "<a href=""" & Blog.googlePlusLink & """ target=""_blank""><img Class=""blogSidebarSocialLogo"" src=""/blogs/GooglePlus.jpg"" width=""32"" height=""32""></a>"
                            ElseIf Blog.allowGooglePlusLink Then
                                If cp.User.IsAdmin Then
                                    sidebarHtml &= "<div Class=""blogAdminWarning""><h2>Administrator</h2><p>Add a GooglePlus link For this blog, Or disable the Allow Google Plus Sidebar checkbox.</p></div>"
                                End If
                            End If
                            If sidebarHtml <> "" Then
                                If Blog.followUsCaption = "" Then
                                    Blog.followUsCaption = "Follow Us"
                                End If
                                sidebarCell.Load(cellTemplate)
                                Call sidebarCell.SetInner(".blogSidebarCellHeadline", Blog.followUsCaption)
                                Call sidebarCell.SetOuter(".blogSidebarCellCopy", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellInput", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellButton", "")
                                Call sidebarCell.SetInner(".blogSidebarCellInputCaption", sidebarHtml)
                                cellList &= vbCrLf & vbTab & "<div id=""blogSidebarSocialCell"">" & sidebarCell.GetHtml() & "</div>"
                                sidebarCnt += 1
                            End If
                        End If
                        '
                        If Blog.allowSearch Then
                            '
                            ' Search 
                            Dim formInput As String
                            formInput = cp.Html.InputText("keywordList", cp.Doc.GetText("keywordList"))
                            formInput += cp.Html.Hidden("formid", "120")
                            formInput += cp.Html.Hidden("sourceformid", "120")
                            formInput += cp.Html.Hidden("button", " Search Blogs ")
                            formInput = cp.Html.Form(formInput, "", "blogSidebarSearchForm", "blogSidebarSearchForm")
                            sidebarCell.Load(cellTemplate)
                            Call sidebarCell.SetInner(".blogSidebarCellHeadline", "Search")
                            Call sidebarCell.SetOuter(".blogSidebarCellCopy", "")
                            Call sidebarCell.SetOuter(".blogSidebarCellInputCaption", "")
                            Call sidebarCell.SetInner(".blogSidebarCellInput", formInput)
                            Call sidebarCell.SetInner(".blogSidebarCellButton", "<a href=""#"" id=""blogSidebarSearch"" onclick=""jQuery('#blogSidebarSearchForm').submit();return false;"">Search</a>")
                            cellList &= vbCrLf & vbTab & "<div id=""blogSidebarSearchCell"">" & sidebarCell.GetHtml() & "</div>"
                            sidebarCnt += 1
                        End If
                        '
                        If Blog.allowArchiveList Then
                            '
                            ' Archive List
                            Dim sideBar_ArchiveList As String = ""
                            '
                            ' create the article list now - if only the current month, turn if off before setting allowListSidebar
                            ''
                            'Dim blogListQs As String = cp.Doc.RefreshQueryString()
                            'blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameSourceFormID, "")
                            'blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameFormID, "")
                            'blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameBlogCategoryID, "")
                            'blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameBlogEntryID, "")
                            Dim blogListLink As String = cp.Content.GetLinkAliasByPageID(cp.Doc.PageId, app.blogListLink, "?" & app.blogListLink)
                            sideBar_ArchiveList = SidebarView.GetSidebarArchiveList(cp, app)
                            If (String.IsNullOrEmpty(sideBar_ArchiveList)) Then
                                Blog.allowArchiveList = False
                            Else
                                sidebarCell.Load(cellTemplate)
                                Call sidebarCell.SetInner(".blogSidebarCellHeadline", "Archives")
                                Call sidebarCell.SetInner(".blogSidebarCellCopy", sideBar_ArchiveList)
                                Call sidebarCell.SetOuter(".blogSidebarCellInputCaption", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellInput", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellButton", "")
                                cellList &= vbCrLf & vbTab & "<div id=""blogSidebarArchiveCell"">" & sidebarCell.GetHtml() & "</div>"
                                sidebarCnt += 1
                            End If
                        End If
                        '
                        If Blog.allowRSSSubscribe Then
                            Dim rssFeed = DbModel.create(Of RSSFeedModel)(cp, Blog.RSSFeedID)
                            '
                            If ((rssFeed Is Nothing) OrElse (rssFeed.rssFilename = "")) Then
                                adminSuggestions &= cp.Html.li("This blog includes an RSS Feed, but no feed has been created. It his persists, please contact the site developer. Disable RSS feeds for this blog to hide this message.")
                            Else
                                '
                                sidebarCell.Load(cellTemplate)
                                Call sidebarCell.SetInner(".blogSidebarCellHeadline", "Subscribe By RSS")
                                Call sidebarCell.SetOuter(".blogSidebarCellCopy", "")
                                'http://aoblog/aoblog/files/rssfeeds/RSSFeed8.xml
                                Call sidebarCell.SetInner(".blogSidebarCellInputCaption", "<a href=""http://" & cp.Site.DomainPrimary & cp.Http.CdnFilePathPrefix & rssFeed.rssFilename & """><img id=""blogSidebarRSSLogo"" src=""/blogs/rss.png"" width=""25"" height=""25"">" & blog.name & " Feed" & "</a>")
                                Call sidebarCell.SetOuter(".blogSidebarCellInput", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellButton", "")
                                cellList &= vbCrLf & vbTab & "<div id=""blogSidebarRSSCell"">" & sidebarCell.GetHtml() & "</div>"
                                sidebarCnt += 1
                            End If
                        End If
                        '
                        If isArticleView Then
                            Dim emtyblogEntryCtaRuleList = DbModel.createList(Of BlogEntryCTARuleModel)(cp, "blogEntryid=" & request.blogEntryId)
                            If emtyblogEntryCtaRuleList.Count > 0 Then

                            Else
                                sidebarCell.Load(cellTemplate)
                                Call sidebarCell.SetInner(".blogSidebarCellHeadline", "")
                                Call sidebarCell.SetInner(".blogSidebarCellCopy", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellInput", "")
                                Call sidebarCell.SetOuter(".blogSidebarCellInputCaption", "")
                                If cp.User.IsAdmin Then
                                    cellList &= vbCrLf & vbTab & "<div class=""aoBlogFooterLink"" style=""color:red;"">Blog Post has no CTA selected</a><br></div>"
                                End If
                            End If
                        End If
                        '
                        If Blog.allowArticleCTA Then
                            '
                            ' Call To action List
                            '
                            Dim cta = cp.Content.GetID("Calls To Action")
                            Dim qs As String = ""
                            sidebarCell.Load(cellTemplate)
                            Call sidebarCell.SetInner(".blogSidebarCellHeadline", "Call to Action")
                            Call sidebarCell.SetInner(".blogSidebarCellCopy", "")
                            Call sidebarCell.SetOuter(".blogSidebarCellInputCaption", "")
                            Call sidebarCell.SetOuter(".blogSidebarCellInput", "")
                            Call sidebarCell.SetOuter(".blogSidebarCellButton", "")
                            qs = "cid=" & cp.Content.GetID("Calls To Action")
                            '
                            If cp.User.IsAdmin Then
                                cellList &= vbCrLf & vbTab & "<div class=""aoBlogFooterLink""><a href=""" & cp.Site.GetText("adminUrl") & "?" & qs & """>Add/Edit Site Call-To-Actions</a></div>"
                            End If
                            sidebarCnt += 1
                        End If
                    End If
                    '
                    layout.SetInner(".blogSidebar", cellList)
                    layout.Append(cp.Html.Hidden("blogId", Blog.id.ToString(), "", "blogId"))
                    If sidebarCnt = 0 Then
                        layout.SetInner(".blogWrapper", layout.GetInner(".blogColumn1"))
                    End If
                    Dim resultHtml As String = layout.GetHtml()
                    '
                    ' -- add legacy blog into sidebar wrapper
                    resultHtml = resultHtml.Replace("{{legacyBlog}}", legacyBlogBody)
                    '
                    ' -- add admin hint to the bottom
                    If adminSuggestions <> "" And cp.User.IsAdmin() Then
                        resultHtml &= "<div class=""ccHintWrapper""><div class=""ccHintWrapperContent""><h2>Administrator</h2><ul>" & adminSuggestions & "</ul></div></div>"
                    End If
                    Return resultHtml
                End Using
            Catch ex As Exception
                cp.Site.ErrorReport(ex)
                Throw
            End Try
        End Function
        '
        '====================================================================================
        '
        Public Shared Function GetSidebarArchiveList(cp As CPBaseClass, app As ApplicationEnvironmentModel) As String
            Dim returnHtml As String = ""
            Try
                Dim cs As CPCSBaseClass = cp.CSNew()
                Dim ArchiveMonth As Integer
                Dim ArchiveYear As Integer
                Dim NameOfMonth As String
                Dim qs As String
                Dim SQL As String
                '
                SQL = "SELECT distinct Month(DateAdded) as ArchiveMonth, year(dateadded) as ArchiveYear " _
                    & " From ccBlogCopy" _
                    & " Where (ContentControlID = " & cp.Content.GetID(cnBlogEntries) & ") And (Active <> 0)" _
                    & " AND (BlogID=" & app.blog.id & ")" _
                    & " ORDER BY year(dateadded) desc, Month(DateAdded) desc"
                If cs.OpenSQL(SQL) Then
                    qs = app.blogListLink
                    qs = cp.Utils.ModifyQueryString(qs, RequestNameFormID, FormBlogArchivedBlogs.ToString())
                    Do While cs.OK
                        ArchiveMonth = cs.GetInteger("ArchiveMonth")
                        ArchiveYear = cs.GetInteger("ArchiveYear")
                        NameOfMonth = MonthName(ArchiveMonth)
                        qs = cp.Utils.ModifyQueryString(qs, RequestNameArchiveMonth, CStr(ArchiveMonth))
                        qs = cp.Utils.ModifyQueryString(qs, RequestNameArchiveYear, CStr(ArchiveYear))
                        returnHtml = returnHtml & vbCrLf & vbTab & vbTab & "<li class=""aoBlogArchiveLink""><a href=""?" & qs & """>" & NameOfMonth & "&nbsp;" & ArchiveYear & "</a></li>"
                        Call cs.GoNext()
                    Loop
                End If
                Call cs.Close()
                If returnHtml <> "" Then
                    returnHtml = "" _
                        & vbCrLf & vbTab & "<ul class=""aoBlogArchiveLinkList"">" _
                        & returnHtml _
                        & vbCrLf & vbTab & "</ul>"
                End If
                '
            Catch ex As Exception
                cp.Site.ErrorReport(ex)
            End Try
            Return returnHtml
        End Function
        '
    End Class
End Namespace

File: source/aoBlogs2/Views/SidebarView.cs

using Contensive.Models.Db;
using System;
using Contensive.Addons.Blog.Models;
using Contensive.Addons.Blog.Models.View;
using Contensive.BaseClasses;
using Microsoft.VisualBasic;

namespace Contensive.Addons.Blog.Views {
    // 
    public class SidebarView {
        // 
        public static string getSidebarView(CPBaseClass cp, ApplicationEnvironmentModel app, RequestModel request, string legacyBlogBody) {
            try {
                var blog = app.blog;
                string cellList = "";
                using (var sidebarCell = cp.BlockNew()) {
                    var layout = cp.BlockNew();
                    layout.OpenLayout(constants.BlogListLayout);
                    sidebarCell.Load(layout.GetOuter(".blogSidebarCellWrap"));
                    string cellTemplate = sidebarCell.GetHtml();
                    string adminSuggestions = "";
                    int sidebarCnt = 0;
                    // 
                    // -- wrap the blog body in the wrapper (sidebar, footer, etc)
                    bool isArticleView = request.blogEntryId > 0;
                    bool allowListSidebar = blog.allowEmailSubscribe | blog.allowFacebookLink | blog.allowGooglePlusLink | blog.allowGooglePlusLink | blog.allowRSSSubscribe | blog.allowTwitterLink | blog.allowArchiveList | blog.allowSearch;
                    bool allowArticleSidebar = allowListSidebar | blog.allowArticleCTA;
                    int dstFormId = cp.Doc.GetInteger(constants.rnFormID);
                    bool allowSidebar = !(dstFormId == constants.FormBlogEntryEditor) & (isArticleView & allowArticleSidebar | !isArticleView & allowListSidebar);
                    // 
                    // Sidebar
                    if (allowSidebar) {
                        if (blog.allowArticleCTA & isArticleView) {
                            // 
                            // CTA cells
                            var blogEntryCtaRuleList = DbBaseModel.createList<BlogEntryCTARuleModel>(cp, "blogEntryid=" + request.blogEntryId);
                            foreach (var rule in blogEntryCtaRuleList) {
                                var cta = DbBaseModel.create<CallsToActionModel>(cp, rule.calltoactionid);
                                if (cta is not null) {
                                    sidebarCell.Load(cellTemplate);
                                    sidebarCell.SetInner(".blogSidebarCellHeadline", cta.headline + "<br>");
                                    sidebarCell.SetInner(".blogSidebarCellCopy", cta.brief);
                                    sidebarCell.SetOuter(".blogSidebarCellInput", "");
                                    sidebarCell.SetOuter(".blogSidebarCellInputCaption", "");
                                    if (string.IsNullOrEmpty(cta.name) | string.IsNullOrEmpty(cta.link)) {
                                        sidebarCell.SetOuter(".blogSidebarCellButton", "");
                                    }
                                    else {
                                        sidebarCell.SetInner(".blogSidebarCellButton", "<a target=\"_blank\" href=\"" + cta.link + "\">" + cta.name + "</a>");
                                    }
                                    cellList += Constants.vbCrLf + Constants.vbTab + sidebarCell.GetHtml();
                                    sidebarCnt += 1;
                                }
                            }
                        }
                        // 
                        if (blog.allowEmailSubscribe) {
                            // 
                            // Subscribe by email
                            // 
                            bool subscribed = cp.Visit.GetBoolean("EmailSubscribed-Blog" + blog.id + "-user" + cp.User.Id);
                            if (!subscribed) {
                                subscribed = cp.User.IsInGroup(blog.emailSubscribeGroupId.ToString());
                            }
                            sidebarCell.Load(cellTemplate);
                            sidebarCell.SetInner(".blogSidebarCellHeadline", "Subscribe By Email");
                            sidebarCell.SetOuter(".blogSidebarCellCopy", "");
                            if (subscribed) {
                                sidebarCell.SetInner(".blogSidebarCellInputCaption", "You are subscribed to this blog.");
                                sidebarCell.SetOuter(".blogSidebarCellInput", "");
                                sidebarCell.SetOuter(".blogSidebarCellButton", "");
                            }

                            else {
                                sidebarCell.SetInner(".blogSidebarCellInputCaption", "Email*");
                                sidebarCell.SetInner(".blogSidebarCellInput", "<input type=\"text\" id=\"blogSubscribeEmail\" name=\"email\" value=\"" + cp.User.Email + "\">");
                                sidebarCell.SetInner(".blogSidebarCellButton a", "Subscribe");
                                sidebarCell.SetInner(".blogSidebarCellButton", "<a href=\"#\" id=\"blogSidebarEmailSubscribe\">Subscribe</a>");
                            }
                            cellList += Constants.vbCrLf + Constants.vbTab + "<div id=\"blogSidebarEmailCell\">" + sidebarCell.GetHtml() + "</div>";
                            sidebarCnt += 1;
                        }
                        // 
                        if (blog.allowFacebookLink | blog.allowGooglePlusLink | blog.allowTwitterLink) {
                            // 
                            // Social Links
                            // Dim sidebarHtml As String = sidebarCell.GetHtml()
                            string sidebarHtml = "";
                            if (blog.allowFacebookLink & !string.IsNullOrEmpty(blog.facebookLink)) {
                                sidebarHtml += "<a href=\"" + blog.facebookLink + "\" target=\"_blank\"><img Class=\"blogSidebarSocialLogo\" src=\"/blogs/facebook.jpg\" width=\"32\" height=\"32\"></a>";
                            }
                            else if (blog.allowFacebookLink) {
                                if (cp.User.IsAdmin) {
                                    sidebarHtml += "<div Class=\"blogAdminWarning\"><h2>Administrator</h2><p>Add a facebook link For this blog, Or disable the Allow Facebook Sidebar checkbox.</p></div>";
                                }
                            }
                            if (blog.allowTwitterLink & !string.IsNullOrEmpty(blog.twitterLink)) {
                                sidebarHtml += "<a href=\"" + blog.twitterLink + "\" target=\"_blank\"><img Class=\"blogSidebarSocialLogo\" src=\"/blogs/twitter.jpg\" width=\"32\" height=\"32\"></a>";
                            }
                            else if (blog.allowTwitterLink) {
                                if (cp.User.IsAdmin) {
                                    sidebarHtml += "<div Class=\"blogAdminWarning\"><h2>Administrator</h2><p>Add a twitter link For this blog, Or disable the Allow Twitter Sidebar checkbox.</p></div>";
                                }
                            }
                            if (blog.allowTwitterLink & !string.IsNullOrEmpty(blog.googlePlusLink)) {
                                sidebarHtml += "<a href=\"" + blog.googlePlusLink + "\" target=\"_blank\"><img Class=\"blogSidebarSocialLogo\" src=\"/blogs/GooglePlus.jpg\" width=\"32\" height=\"32\"></a>";
                            }
                            else if (blog.allowGooglePlusLink) {
                                if (cp.User.IsAdmin) {
                                    sidebarHtml += "<div Class=\"blogAdminWarning\"><h2>Administrator</h2><p>Add a GooglePlus link For this blog, Or disable the Allow Google Plus Sidebar checkbox.</p></div>";
                                }
                            }
                            if (!string.IsNullOrEmpty(sidebarHtml)) {
                                if (string.IsNullOrEmpty(blog.followUsCaption)) {
                                    blog.followUsCaption = "Follow Us";
                                }
                                sidebarCell.Load(cellTemplate);
                                sidebarCell.SetInner(".blogSidebarCellHeadline", blog.followUsCaption);
                                sidebarCell.SetOuter(".blogSidebarCellCopy", "");
                                sidebarCell.SetOuter(".blogSidebarCellInput", "");
                                sidebarCell.SetOuter(".blogSidebarCellButton", "");
                                sidebarCell.SetInner(".blogSidebarCellInputCaption", sidebarHtml);
                                cellList += Constants.vbCrLf + Constants.vbTab + "<div id=\"blogSidebarSocialCell\">" + sidebarCell.GetHtml() + "</div>";
                                sidebarCnt += 1;
                            }
                        }
                        // 
                        if (blog.allowSearch) {
                            // 
                            // Search 
                            string formInput;
                            formInput = cp.Html.InputText("keywordList", cp.Doc.GetText("keywordList"));
                            formInput += cp.Html.Hidden("formid", "120");
                            formInput += cp.Html.Hidden("sourceformid", "120");
                            formInput += cp.Html.Hidden("button", " Search Blogs ");
                            formInput = cp.Html.Form(formInput, "", "blogSidebarSearchForm", "blogSidebarSearchForm");
                            sidebarCell.Load(cellTemplate);
                            sidebarCell.SetInner(".blogSidebarCellHeadline", "Search");
                            sidebarCell.SetOuter(".blogSidebarCellCopy", "");
                            sidebarCell.SetOuter(".blogSidebarCellInputCaption", "");
                            sidebarCell.SetInner(".blogSidebarCellInput", formInput);
                            sidebarCell.SetInner(".blogSidebarCellButton", "<a href=\"#\" id=\"blogSidebarSearch\" onclick=\"jQuery('#blogSidebarSearchForm').submit();return false;\">Search</a>");
                            cellList += Constants.vbCrLf + Constants.vbTab + "<div id=\"blogSidebarSearchCell\">" + sidebarCell.GetHtml() + "</div>";
                            sidebarCnt += 1;
                        }
                        // 
                        if (blog.allowArchiveList) {
                            // 
                            // Archive List
                            string sideBar_ArchiveList = "";
                            // 
                            // create the article list now - if only the current month, turn if off before setting allowListSidebar
                            // '
                            // Dim blogListQs As String = cp.Doc.RefreshQueryString()
                            // blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameSourceFormID, "")
                            // blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameFormID, "")
                            // blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameBlogCategoryID, "")
                            // blogListQs = cp.Utils.ModifyQueryString(blogListQs, RequestNameBlogEntryID, "")
                            string blogListLink = cp.Content.GetLinkAliasByPageID(cp.Doc.PageId, app.blogBaseLink, "?" + app.blogBaseLink);
                            sideBar_ArchiveList = GetSidebarArchiveList(cp, app);
                            if (string.IsNullOrEmpty(sideBar_ArchiveList)) {
                                blog.allowArchiveList = false;
                            }
                            else {
                                sidebarCell.Load(cellTemplate);
                                sidebarCell.SetInner(".blogSidebarCellHeadline", "Archives");
                                sidebarCell.SetInner(".blogSidebarCellCopy", sideBar_ArchiveList);
                                sidebarCell.SetOuter(".blogSidebarCellInputCaption", "");
                                sidebarCell.SetOuter(".blogSidebarCellInput", "");
                                sidebarCell.SetOuter(".blogSidebarCellButton", "");
                                cellList += Constants.vbCrLf + Constants.vbTab + "<div id=\"blogSidebarArchiveCell\">" + sidebarCell.GetHtml() + "</div>";
                                sidebarCnt += 1;
                            }
                        }
                        // 
                        if (blog.allowRSSSubscribe) {
                            var rssFeed = DbBaseModel.create<RSSFeedModel>(cp, blog.RSSFeedID);
                            // 
                            if (rssFeed is null || string.IsNullOrEmpty(rssFeed.rssFilename)) {
                                adminSuggestions += cp.Html.li("This blog includes an RSS Feed, but no feed has been created. It his persists, please contact the site developer. Disable RSS feeds for this blog to hide this message.");
                            }
                            else {
                                // 
                                sidebarCell.Load(cellTemplate);
                                sidebarCell.SetInner(".blogSidebarCellHeadline", "Subscribe By RSS");
                                sidebarCell.SetOuter(".blogSidebarCellCopy", "");
                                // http://aoblog/aoblog/files/rssfeeds/RSSFeed8.xml
                                sidebarCell.SetInner(".blogSidebarCellInputCaption", "<a href=\"http://" + cp.Site.DomainPrimary + cp.Http.CdnFilePathPrefix + rssFeed.rssFilename + "\"><img id=\"blogSidebarRSSLogo\" src=\"/blogs/rss.png\" width=\"25\" height=\"25\">" + blog.name + " Feed" + "</a>");
                                sidebarCell.SetOuter(".blogSidebarCellInput", "");
                                sidebarCell.SetOuter(".blogSidebarCellButton", "");
                                cellList += Constants.vbCrLf + Constants.vbTab + "<div id=\"blogSidebarRSSCell\">" + sidebarCell.GetHtml() + "</div>";
                                sidebarCnt += 1;
                            }
                        }
                        // 
                        if (isArticleView) {
                            var emtyblogEntryCtaRuleList = DbBaseModel.createList<BlogEntryCTARuleModel>(cp, "blogEntryid=" + request.blogEntryId);
                            if (emtyblogEntryCtaRuleList.Count > 0) {
                            }

                            else {
                                sidebarCell.Load(cellTemplate);
                                sidebarCell.SetInner(".blogSidebarCellHeadline", "");
                                sidebarCell.SetInner(".blogSidebarCellCopy", "");
                                sidebarCell.SetOuter(".blogSidebarCellInput", "");
                                sidebarCell.SetOuter(".blogSidebarCellInputCaption", "");
                                if (cp.User.IsAdmin) {
                                    cellList += Constants.vbCrLf + Constants.vbTab + "<div class=\"aoBlogFooterLink\" style=\"color:red;\">Blog Post has no CTA selected</a><br></div>";
                                }
                            }
                        }
                        // 
                        if (blog.allowArticleCTA) {
                            // 
                            // Call To action List
                            // 
                            int cta = cp.Content.GetID("Calls To Action");
                            string qs = "";
                            sidebarCell.Load(cellTemplate);
                            sidebarCell.SetInner(".blogSidebarCellHeadline", "Call to Action");
                            sidebarCell.SetInner(".blogSidebarCellCopy", "");
                            sidebarCell.SetOuter(".blogSidebarCellInputCaption", "");
                            sidebarCell.SetOuter(".blogSidebarCellInput", "");
                            sidebarCell.SetOuter(".blogSidebarCellButton", "");
                            qs = "cid=" + cp.Content.GetID("Calls To Action");
                            // 
                            if (cp.User.IsAdmin) {
                                cellList += Constants.vbCrLf + Constants.vbTab + "<div class=\"aoBlogFooterLink\"><a href=\"" + cp.Site.GetText("adminUrl") + "?" + qs + "\">Add/Edit Site Call-To-Actions</a></div>";
                            }
                            sidebarCnt += 1;
                        }
                    }
                    // 
                    layout.SetInner(".blogSidebar", cellList);
                    layout.Append(cp.Html.Hidden("blogId", blog.id.ToString(), "", "blogId"));
                    if (sidebarCnt == 0) {
                        layout.SetInner(".blogWrapper", layout.GetInner(".blogColumn1"));
                    }
                    string resultHtml = layout.GetHtml();
                    // 
                    // -- add legacy blog into sidebar wrapper
                    resultHtml = resultHtml.Replace("{{legacyBlog}}", legacyBlogBody);
                    // 
                    // -- add admin hint to the bottom
                    if (!string.IsNullOrEmpty(adminSuggestions) & cp.User.IsAdmin) {
                        resultHtml += "<div class=\"ccHintWrapper\"><div class=\"ccHintWrapperContent\"><h2>Administrator</h2><ul>" + adminSuggestions + "</ul></div></div>";
                    }
                    return resultHtml;
                }
            }
            catch (Exception ex) {
                cp.Site.ErrorReport(ex);
                throw;
            }
        }
        // 
        // ====================================================================================
        // 
        public static string GetSidebarArchiveList(CPBaseClass cp, ApplicationEnvironmentModel app) {
            string returnHtml = "";
            try {
                var cs = cp.CSNew();
                int ArchiveMonth;
                int ArchiveYear;
                string NameOfMonth;
                string SQL;
                // 
                SQL = "SELECT distinct Month(dateAdded) as ArchiveMonth, year(dateAdded) as ArchiveYear " + " From ccBlogCopy" + " Where (ContentControlID = " + cp.Content.GetID(constants.cnBlogEntries) + ") And (Active <> 0)" + " AND (BlogID=" + app.blog.id + ")" + " ORDER BY year(dateAdded) desc, Month(dateAdded) desc";



                if (cs.OpenSQL(SQL)) {
                    //qs = app.blogBaseLink;
                    string qs = cp.Utils.ModifyQueryString("", constants.rnFormID, constants.FormBlogArchivedBlogs.ToString());
                    while (cs.OK()) {
                        ArchiveMonth = cs.GetInteger("ArchiveMonth");
                        ArchiveYear = cs.GetInteger("ArchiveYear");
                        NameOfMonth = DateAndTime.MonthName(ArchiveMonth);
                        qs = cp.Utils.ModifyQueryString(qs, constants.RequestNameArchiveMonth, ArchiveMonth.ToString());
                        qs = cp.Utils.ModifyQueryString(qs, constants.RequestNameArchiveYear, ArchiveYear.ToString());
                        returnHtml = returnHtml + Constants.vbCrLf + Constants.vbTab + Constants.vbTab + "<li class=\"aoBlogArchiveLink\"><a href=\"" + app.blogBaseLink + "?" + qs + "\">" + NameOfMonth + "&nbsp;" + ArchiveYear + "</a></li>";
                        cs.GoNext();
                    }
                }
                cs.Close();
                if (!string.IsNullOrEmpty(returnHtml)) {
                    returnHtml = "" + Constants.vbCrLf + Constants.vbTab + "<ul class=\"aoBlogArchiveLinkList\">" + returnHtml + Constants.vbCrLf + Constants.vbTab + "</ul>";


                }
            }
            // 
            catch (Exception ex) {
                cp.Site.ErrorReport(ex);
            }
            return returnHtml;
        }
        // 
    }
}